Movie Challenge/Movie Challenge/Controllers/Detail/DetailViewController.swift

```diff
@@ -0,0 +1,153 @@
+                setButtonState()
+                
+                print("Removeu")
+            }catch let error{
+                print("Erro ao deletar o filme", error)
+            }
+            
+        }else{
+            do{
+                try MovieRepository.shared().save(movie: movie)
+                favorite = true
+                setButtonState()
+                
+                print("Salvou")
+            }catch let error{
+                print("Erro ao salvar o filme", error)
+            }
+        }
+    }
+    
+    private func setPoster(poster: UIImage){
+        DispatchQueue.main.async() {
+            self.posterImageView.image = poster
+            self.posterImageView.setNeedsDisplay()
+        }
+    }
+    
+    private func setButtonState(){
+        if(favorite){
+            DispatchQueue.main.async(){
+                self.favoriteButtonView.backgroundColor = #colorLiteral(red: 0.9254902005, green: 0.2352941185, blue: 0.1019607857, alpha: 1)
+                self.favoriteButtonView.setTitle("Remover", for: UIControlState.normal)
+            }
+        }else{
+            DispatchQueue.main.async(){
+                self.favoriteButtonView.backgroundColor = #colorLiteral(red: 1, green: 0.6116010603, blue: 0.006196474039, alpha: 1)
+                self.favoriteButtonView.setTitle("Favoritar", for: UIControlState.normal)
+            }
+        }
+    }
+    
+    override func viewDidLoad() {
+        categoryCollectionView.delegate = self
+        categoryCollectionView.dataSource = self
+        
+        self.movieId = movieId ?? 76341
+        
+        do{
+            self.movie = try MovieRepository.shared().getOne(by: movieId)
+            self.favorite = true
+            self.setButtonState()
+            self.fillFields()
+        }catch{
+            self.favorite = false
+            
+            _ = MovieService.shared().getMovieDetail(id: movieId){ movie, response, error in
+                self.movie = movie
+                self.setButtonState()
+                self.fillFields()
+            }
+        }
+    }
+    
+    public func setUp(movieId: Int?){
+        //TODO MAD MAX CODE
+        self.movieId = movieId ?? 76341
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        guard let genres = movie.genres else { return 0 }
+        
+        return genres.count
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "CategoryCell", for: indexPath) as! CategoryCollectionViewCell
+        
+        guard let categoryList = movie.genres else{ return UICollectionViewCell() }
+        cell.setUp(name: categoryList[indexPath.row].name)
+        
+        return cell
+    }
+    
+    func fillFields(){
+        DispatchQueue.main.async(){
```

Comment:
One more time, verify if you really need to force your code to main thread, but every time you need to use the main thread instead do not dispatch inside function but call the function inside the dispatch.